### PR TITLE
[wasm] Improve AOT builds

### PR DESCRIPTION
**Pull request by radical**
- Fix incremental AOT builds:
   If you build with AOT, and then build again, then the build breaks. The main reason is that we cil-strip the runtime 
   pack assemblies in-place, which breaks the subsequent build.

- This also fixes wasm targets to work better with blazor workloads.
- Fix targets to work better on windows
- Add a new `RunWithEmSdkEnv` task

### CODE REVIEW

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.Sdk.props
Comment (pranavkm) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.Sdk.props:4]:
```suggestion
    <MonoAotCrossCompilerPath Condition="$([MSBuild]::IsOsPlatform('Windows'))">$(MonoAotCrossCompilerPath).exe</MonoAotCrossCompilerPath>
```

File: src/mono/wasm/build/WasmApp.targets
Comment (pranavkm) [inline src/mono/wasm/build/WasmApp.targets:86]: Are you able to use `$(TargetFramework)` here instead? Avoids having to keep this file updated between major releases.

File: src/mono/wasm/build/WasmApp.targets
Comment (pranavkm) [inline src/mono/wasm/build/WasmApp.targets:235]:
```suggestion
      <_EmsdkEnvSourceCommand Condition="$([MSBuild]::IsOSPlatform('WINDOWS'))">$(EMSDK_PATH)\emsdk_env.bat &gt; nul 2&gt;&amp;1 </_EmsdkEnvSourceCommand>
```

File: src/mono/wasm/build/WasmApp.targets
Comment (pranavkm) [inline src/mono/wasm/build/WasmApp.targets:246]: Could you file a tracking issue to determine if shelling out to bash could be removed? It adds an extra dependency that might not be available in all environments.

File: src/mono/wasm/build/WasmApp.targets
Comment (pranavkm) [inline src/mono/wasm/build/WasmApp.targets:248]: You could use ConsoleToMsBuild. https://stackoverflow.com/a/16804205

File: src/mono/wasm/build/WasmApp.targets
Comment (pranavkm) [inline src/mono/wasm/build/WasmApp.targets:362]: Should the SDK start using this item group?

File: src/mono/wasm/build/WasmApp.targets
Comment (pranavkm) [inline src/mono/wasm/build/WasmApp.targets:308]: Does it work without shelling out?

File: src/mono/wasm/build/WasmApp.targets
Comment (radical) [reply src/mono/wasm/build/WasmApp.targets:362]: Yeah, once we merge this

File: src/mono/wasm/build/WasmApp.targets
Comment (radical) [reply src/mono/wasm/build/WasmApp.targets:246]: On non-windows platforms, we would it, I believe, to run `emsdk_env.sh`. What platforms might have it missing, and we would be able to install Emscripten?

File: src/mono/wasm/build/WasmApp.targets
Comment (radical) [reply src/mono/wasm/build/WasmApp.targets:308]: Maybe.. I can run `$EMSDK_PATH/emsdk emcc`, but I don't know what it might require to compile this or other stuff. If somebody familiar with emsdk can say that it's safe to use that, then I can do that. Maybe @kg, @vargaz, or @lewing?

File: src/tasks/WasmAppBuilder/RunWithEmSdkEnv.cs
Comment (pranavkm) [inline src/tasks/WasmAppBuilder/RunWithEmSdkEnv.cs:40]: re
```suggestion
            return !Log.HasLoggedErrors;
```

File: src/tasks/WasmAppBuilder/RunWithEmSdkEnv.cs
Comment (pranavkm) [inline src/tasks/WasmAppBuilder/RunWithEmSdkEnv.cs:12]: 👏🏽

File: src/tasks/WasmAppBuilder/ReadEmccVersion.cs
Comment (kg) [inline src/tasks/WasmAppBuilder/ReadEmccVersion.cs:31]: I don't think we need a special task for this. MSBuild has one built in:
https://docs.microsoft.com/en-us/visualstudio/msbuild/readlinesfromfile-task?view=vs-2019

Comment (lewing) [reply src/tasks/WasmAppBuilder/ReadEmccVersion.cs:31]: yeah you should be able to read the version with something like
```
    <ReadLinesFromFile File="@(VersionFile)">
      <Output TaskParameter="Lines" ItemName="VersionLines" />
    </ReadLinesFromFile>
    <PropertyGroup>
      <EmccVersion>@(VersionLines)</EmccVersion>
    </PropertyGroup>
```

Comment (radical) [reply src/tasks/WasmAppBuilder/ReadEmccVersion.cs:31]: That's what is already there. The problem was reading one line, on windows. The whole output is like:
```
emcc (Emscripten gcc/clang-like replacement) 2.0.11 (6e28e4fa4fa1bc50d58b9ddbbb9603a3cf21ea9e)
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

```

Comment (pranavkm) [reply src/tasks/WasmAppBuilder/ReadEmccVersion.cs:31]: Could you use `StartsWith` instead?

```xml
<Error Condition="!$(ActualVersion.StartsWith($(ExpectedVersion))" Text="..." />
```

that solves this issue.

Comment (pranavkm) [reply src/tasks/WasmAppBuilder/ReadEmccVersion.cs:31]: Unrelated, how willing would emsdk be to change this command output? It's really weird that asking for a version includes licensing information.

Comment (marek-safar) [reply src/tasks/WasmAppBuilder/ReadEmccVersion.cs:31]: We could submit PR

Comment (vargaz) [reply src/tasks/WasmAppBuilder/ReadEmccVersion.cs:31]: The output is probably not designed to be machine parsable. Can't msbuild read the first line only, or extract the first item in @(VersionLines) ?

Comment (pranavkm) [reply src/tasks/WasmAppBuilder/ReadEmccVersion.cs:31]: Something like this might work:

```xml
<EmccVersion>@(VersionLines->Reverse())</EmccVersion>
```

Comment (radical) [reply src/tasks/WasmAppBuilder/ReadEmccVersion.cs:31]: `<EmccVersion Condition="$([System.String]::new('%(Lines.Identity)').Contains('emcc'))">@(Lines)</EmccVersion>`

Comment (lewing) [reply src/tasks/WasmAppBuilder/ReadEmccVersion.cs]: try `"$([System.IO.File]::ReadAllText('$(VersionsFile)')[0])"`

Comment (radical) [reply src/tasks/WasmAppBuilder/ReadEmccVersion.cs]: It doesn't let me call `File.ReadAllLines`. So, now I have:
```xml
    <ItemGroup>
      <_ReversedVersionLines Include="@(_VersionLines->Reverse())" />
    </ItemGroup>
    <PropertyGroup>
      <ActualEmccVersion>%(_ReversedVersionLines.Identity)</ActualEmccVersion>
    </PropertyGroup>
```

I think this is a decent enough solution for now, and it doesn't depend on any specific string/char in the output.

Comment (lewing) [reply src/tasks/WasmAppBuilder/ReadEmccVersion.cs]: I meant ReadAllLines but I see that isn't available :(
The Reverse() call looks best

File: src/mono/wasm/build/WasmApp.targets
Comment (radical) [reply src/mono/wasm/build/WasmApp.targets:88]: `$(TargetFramework)` can be different from what the runtime pack has, right? I'm not sure how I can get the one to be used with the pack.

Comment (pranavkm) [reply src/mono/wasm/build/WasmApp.targets:88]: Not sure I follow. The TFM for the runtime pack is determined by the project's TFM (by the SDK). I haven't seen a case where they would be different.

**Review by steveisok — APPROVED** [review]

**Review by lewing — APPROVED** [review]
This looks good to me, I talked to @radical about some improvements in allowing InterpOnly metadata on the assembly items in AotIterp mode and a couple of naming changes in a follow up but we want to unblock @pranavkm for now.

**Pull request merged**
